### PR TITLE
Add "deactivate callback" command.

### PR DESCRIPTION
**Pull request by torfmaster**
### Pull Request Overview

This pull request adds a "drop callback" and "drop allow" functionality in order to be able to deactivate callbacks. Not beeing able to deactivate a callback forces user data pointers passed to the callback to have unlimited lifetime.

Callbacks are dropped by subscribing a callback with a null function pointer. Allows are dropped by allowing a null slice.

### Testing Strategy
Positive cases:
We tested the new function in the temperature driver by setting a callback and deleting it before running "subscribe". The original callback is not triggered anymore.
We tested the new allow by allowing a proper string to the console and then allowing a null pointer. The string is not printed anymore but the driver is still able to print strings.

Regression testing: 
Temperature and console drivers have been tested directly and are working. The rot13 service and client are still working - this is in my opinion the most subtle point.

### TODO or Help Wanted
I'd be happy about a code review as well as more regression testing.

### Documentation Updated

- [x] Kernel: The relevant files in `/docs` have been updated or no updates are required.
- [x] Userland: The application README has been added, updated, or no updates are required.

### Formatting

- [x] `make formatall` has been run.

### CODE REVIEW

**Comment by brghena** [comment]
Good catch! It's definitely a problem that we don't have a solution for this. It applies to all `subscribe` and `allow` calls too, not just this temperature driver example.

However, I feel like the alternative "passing null as a callback" solution may be the correct one. Coming from C, the null solution feels like the natural solution to me. For languages like Rust, it feels like it could be accomplished with a `None`. Thoughts?

**Comment by torfmaster** [comment]
I think changing "subscribe" in the Driver trait to accept Option<Callback> and mapping 0 as usize to None would be an option.

**Comment by phil-levis** [comment]
I agree that passing a None to subscribe is probably the right approach, rather than a command. This is important for when there are multiple callbacks. From the C side, I think rather than NULL it might be better to follow the model of sigaction(2), and have a value that denotes "no callback". This could in practice be NULL, but it would be better to be defined to be NO_CALLBACK or something similar (like SIGN_IGN).

**Comment by alevy** [comment]
I agree with @phil-levis that using NULL for the callback to cancel is probably better. We can do this by implementing `Callback#schedule` to check if the function pointer is valid. That shouldn't require changes to any drivers that already just replace the existing callback if a new one is received.

**Comment by torfmaster** [comment]
I propose to change the "Driver" trait to accept an Option<Callback>. For most drivers this would be more or less be a refactoring as the Callback is stored internally as an Option<Callback>. Moreover, interpreting the function pointer in sched.rs makes more sense to me. Any thoughts?

**Comment by ppannuto** [comment]
I agree with the discussion here, to summarize the changes:

  - in-kernel interfaces:
      - `Driver::subscribe` should take an `Option<Callback>` instead of just a Callback
      - `Driver::allow` should take an `Option<AppSlice<Shared, u8>>` instead of just an AppSlice
  - Kernel ABI (recall at this level every parameter is a type-less integer)
      - subscribe shall now interpret parameter `r2` as `None` if 0 instead of returning `EINVAL`
      - allow shall interpret parameter `r2` as `None` if 0 instead of returning `EINVAL`

@torfmaster are you willing to implement these?

**Comment by torfmaster** [comment]
Yes, I'd like to implement that.

**Comment by torfmaster** [comment]
Okay, I implemented a first draft. Next, I'll be testing and refactoring.

**Comment by torfmaster** [comment]
Okay, I'm done and I'm interested in your comments.

File: userland/libtock/tock.h [hunk 1]
```diff
@@ -59,6 +59,9 @@ bool driver_exists(uint32_t driver);
 #define TOCK_EUNINSTALLED -12
 #define TOCK_ENOACK       -13
 
+#define DEACTIVATE_CALLBACK 0
+#define REVOKE_ALLOW 	    0
```
Comment (ppannuto) [inline userland/libtock/tock.h]: I think these should be prefixed with `TOCK_` to be consistent

**Comment by torfmaster** [comment]
Okay, thanks. I changed the constants.

**Comment by alevy** [comment]
@phil-levis, are you happy with these changes (if so can you switch your review to approve?). Other @helena-project/core-team that have not approved yet, let's consider the 1-week clock as ticking.

File: userland/libtock/tock.h [hunk 2]
```diff
@@ -59,6 +59,9 @@ bool driver_exists(uint32_t driver);
 #define TOCK_EUNINSTALLED -12
 #define TOCK_ENOACK       -13
 
+#define TOCK_DEACTIVATE_CALLBACK    0
+#define TOCK_REVOKE_ALLOW           0
```
Comment (bradjc) [inline userland/libtock/tock.h]: A comment here describing when and how these should be used would be nice.

**Comment by torfmaster** [comment]
Okay, I added a comment to tock.h. I hope the travis build won't crash this time.

**Pull request merged**